'B.W specifiable in %' functionality

ndev2 opened the pull request.

milesdevis committed:
'B.W specifiable in %' functionality with updated man entries. Style issues fixed. Changes fixed, checkpatch.pl checked.

cjdoucette commented:
I'm still trying to see if there's a way to avoid changing the function signatures of `parse_qopt()` and `parse_copt()`, but in the meantime you can do two things: (1) update the manual entries (in the `man` directory) for each component of `tc` that can now specify bandwidth as a percentage, and (2) describe the steps needed to move this code upstream in the iproute2 repository.

ndev2 commented:
Will do, I'll try to get these done asap. Thanks.

ndev2 commented:
HI Cody, I've updated the man entries (however tc-gred does not seem to have a man entry at all, but we've mad changes to gred). I've added our functionality as a note in each man entry. 

To move the code upstream, we need to submit it as a patch to the netdev mailing list (I verified this, iproute2 changes are to be mailed to the same mailing list as netdev) and wait for the review of the patch.

Hope the code is good, do let me know if any changes needed. Thanks.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+	int ret;
+	double d, r;
+
+	if (read_prop(dev, "speed", &value))
```
On my Linux system, reading the "speed" property doesn't work for all interfaces. For example, it doesn't work for my wireless interface or for an Ethernet bridge. Can you find a better way to get the device's speed limit?

cjdoucette replied:
There's also no guarantee that the user gave an argument for `dev` when invoking `tc`, so you need to check to make sure it's not `NULL` before trying to use it.

ndev2 replied:
Hi Cody, I've tried searching for better ways but it seems the only other option to find the speed for any interface is to make an `ioctl` call. Would that be ok?

ndev2 replied:
Hi Cody, for wireless interfaces, there is a different ioctl call but for all other wired interfaces - including bridges, which assuming is dependent on the speed of the interface that is connected to the bridge (and that has to be wired) - there is the `ethtool` ioctl call which can find the device speed (they have different ioctl calls because of the different device drivers being used for wired vs wireless connections). But given only the device name from the user we'd have to assume whether it is a wired or wireless connection. Given this, I was thinking the approach would be to search for the device name in the wired device list using `ethtool` ioctl call and if that call fails, search for the device in the wireless devices list, and if that fails, return `no such device found` .

ndev2 replied:
Because this way we're not assuming whether the interface is wired or wireless but we'd still be searching for the right interface because every interface name should be unique.

ndev2 replied:
Let me know if this approach is ok. As of right now `tc` handles calls only pertaining to wired devices (and that's why I guess an error for wireless connections showed up), so the functionality for handling wireless devices needs its appropriate low level functionality to be added into `tc`, before the ioctl call can be made. So do let me know. Thanks.

cjdoucette replied:
OK, thanks for researching that. Let's assume that the interface is wired, and keep the way that you're doing it for now. But if we're unable to read a valid speed, make sure an error is returned. I can't tell whether `read_prop()` will return an error if given the name of a wireless device while looking for the "speed" property, so include that in your testing to make sure. It would be great for `get_percent_rate()` to print an error message in this case too.

ndev2 replied:
Ok, will do. At the moment,`read_prop()` does return an error but `get_precent_rate()` does not, will add that in. Thanks.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+	}
+
+	strncpy(percent, str, strlen(str) -1);
+	ret = sscanf(percent, "%lf", &d);
```
Actually, instead of trying to check whether `str` ends with a `%` and copying everything except the `%` into the variable `percent`, why not just make this statement:

    ret = sscanf(percent, "%lf%%", &d);

Then `sscanf()` will be expecting a floating point number followed by a percentage character.

cjdoucette replied:
Then we can get rid of the string `percent`, pointer `pos`, and the `if` statement  above.

ndev2 replied:
Done. Works good.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+	strncpy(percent, str, strlen(str) -1);
+	ret = sscanf(percent, "%lf", &d);
+	if (ret != 1) {
+		fprintf(stderr, "Specified rate value is malformed \n");
```
Replace the error statement with this one:

    fprintf(stderr, "Specified rate value could not be read or is malformed: %s\n", str);

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+	}
+
+	if (d > 100.0 || d < 0.0) {
+		fprintf(stderr, "Invalid rate specified \n");
```
Replace the error message with this one:

    fprintf(stderr, "Invalid rate specified; should be between [0,100]%, but is %s\n", str);

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
 
+int get_percent_rate(unsigned int *rate, const char *str, char *dev)
+{
+	long value;
```
Rename `value` to be whatever unit we are reading for the speed (maybe `max_rate_bytes` or `max_rate_mbytes`, whatever the unit is we're reading here).

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+
+	r = (d*value) / 100.0;
+
+	ret = snprintf(r_str, 20, "%f", r);
```
It's OK to use `%f` here, but you used `%lf` above, so pick just one and use that consistently.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+	char percent[strlen(dev) - 1], r_str[20];
+	const char *pos;
+	int ret;
+	double d, r;
```
Let's make these variables more readable: make `d` into `perc` and `r` into `rate_bytes` (or whatever the unit ends up being).

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+		return -1;
+	}
+
+	r = (d*value) / 100.0;
```
Add spaces around the `*` operator.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+		return -1;
+	}
+
+	return get_rate(rate, r_str);
```
Don't you need to add a unit measurement ("bit", "mbit", etc) in order for `get_rate()` to work?

ndev2 replied:
No, the `sys/class/net/~/speed` (where `read_prop()` reads from) specifies the speed in bits which is the default suffix in `get_rate`  so when the user specifies BW in %, it'll be taken care of, and the usual case is typical

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+
+	ret = snprintf(r_str, 20, "%f", r);
+	if (ret <= 0 || ret >= 20) {
+		fprintf(stderr, "Invalid rate specified \n");
```
Replace this error message with:

    fprintf(stderr, "Unable to parse calculated rate\n");

cjdoucette commented on tc/tc_util.c:
```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
+
+	return get_rate64(rate, r_str);
+
+}
```
This function almost completely duplicates the code of `get_percent_rate()`, so you should find a way of putting the code that calculates the bandwidth rate in its own function.

ndev2 replied:
Done. But now`get_percent_rate` and `get_percent_rate64` only look like wrappers calling two functions inside them. That should be fine right?

ndev2 replied:
Also another thing I wanted to point out is that the rate parse functions `get_rate` and `get_rate64` do not explicity state the error in the given command, they just `return -1` and let the `explain` functions in the qdisc file handle the explanation of what was wrong. We, as of right now, go against that norm and explicitly state what went wrong inside the parse functions. Should we change this (and modify the `explain` functions in the qdisc files), what do you think?

cjdoucette replied:
I'll need to look at what you ultimately did to factor this code out, but it should be OK that these functions now look like wrappers.

As for error logging, I think what we're doing is OK. I see other instances of using `fprintf()` individually (I'm looking in `tc/q_cbq.c` specifically). It looks like functions like `explain()` and `explain1()` suit more narrower purposes.

cjdoucette reviewed (changes requested):
In addition to the comments on the code here, you should do the following:

Instead of adding a `NOTES` section to each manual page, change the `RATES` section of the main `tc(8)` manual page so say that in addition to specifying a specific rate with a unit, you can specify the rate as a percentage of the device's speed.

Since patches to this repository are done through the kernel mailing list, you should read more about how to prepare for and submit patches:

https://www.kernel.org/doc/html/v4.12/process/submitting-patches.html

https://kernelnewbies.org/FirstKernelPatch

https://www.kernel.org/doc/Documentation/networking/netdev-FAQ.txt

In particular, one of the steps from these guides is to run `checkpatch.pl`. You can find this script in the kernel source, in `scripts/checkpatch.pl`. You should run that script against your changes in this repository to make sure that there are no style issues.

ndev2 commented:
Hi Cody, I've made all the changes but some of the `fprintf` statements go over the 80 character limit (and it does show up as a warning when `checkpatch.pl` is run) but that said, many other `fprintf` statements exceed the limit in the codebase. Thanks.

cjdoucette commented:
OK, go ahead and rebase the code so I can take another look. I'll look out for the 80 character limit isue, too.

ndev2 commented:
Ok, will do. But fair warning, I'm still fixing some whitespace issues in the changed files from `checkpatch.pl`. You should be able to see the rebase now.

cjdoucette commented:
Oh OK, take your time and do that, and then rebase after.

ndev2 commented:
Rebased. Hope the code is good, thanks.

cjdoucette commented on man/man8/tc-fq.8:
```diff
@@ -74,6 +74,7 @@ red-black tree for efficient collision sorting.
 Default: 1024.
 .SS [no]pacing
 Enable or disable flow pacing. Default is enabled.
+
```
You can remove this blank line.

cjdoucette commented on man/man8/tc-tbf.8:
```diff
@@ -115,6 +115,7 @@ allows around 3mbit/s of peakrate, given 1000 byte packets.
 Like the regular burstsize you can also specify a
 .B cell
 size.
+
```
Remove this blank line, too.

cjdoucette commented on man/man8/tc.8:
```diff
@@ -443,7 +443,8 @@ see the man pages for individual qdiscs.
 These parameters accept a floating point number, possibly followed by
-a unit (both SI and IEC units supported).
+either a unit (both SI and IEC units supported), or a '%' character when spec
+ifying the rate as a percentage of the device's speed.
```
Let's make this addition into:

", or a float followed by `%` to specify the rate as a percentage of the device's speed (e.g. 5%, 99.5%)."

cjdoucette commented on tc/tc_util.c:
```diff
@@ -191,6 +191,56 @@ static const struct rate_suffix {
+	strncpy(percent, str, strlen(str) - 1);
+	ret = sscanf(percent, "%lf%%", &perc);
+	if (ret != 1) {
+		fprintf(stderr, "Specified rate value could not be read or is malformed \n");
```
Remove the space before the `\n`.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -191,6 +191,56 @@ static const struct rate_suffix {
+	if (read_prop(dev, "speed", &max_rate_bits))
+		return -1;
+	strncpy(percent, str, strlen(str) - 1);
+	ret = sscanf(percent, "%lf%%", &perc);
```
I don't think we need the `percent` variable at all, right? We can just pass `str` as the first parameter to `sscanf()`, which will parse it into a float followed by a percentage and store the float in `max_rate_bits`.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -191,6 +191,56 @@ static const struct rate_suffix {
+
+	ret = snprintf(rate, 20, "%lf", rate_bits);
+	if (ret <= 0 || ret >= 20) {
+		fprintf(stderr, "Unable to parse calculated rate \n");
```
Remove the space before `\n`.

cjdoucette commented on tc/q_cbq.c:
```diff
@@ -63,7 +63,12 @@ static int cbq_parse_opt(struct qdisc_util *qu, int argc, char **argv, struct nl
 		    matches(*argv, "rate") == 0) {
 			NEXT_ARG();
-			if (get_rate(&r.rate, *argv)) {
+			if (strchr(*argv, '%')) {
```
We still have the issue that the user may have specified a "rate" option for one of the individual qdiscs but did not specify what the device is as part of the main `tc` command. So it's possible that we're reaching this case but `dev` doesn't contain a device name.

You need to check that the device name has been specified. For example, in `tc_class_modify()`, they do the following check (the char array that could hold the device name is called `d` in that function):

    if (d[0])  {
        ...

We need to make sure we check this for all functions that try to calculate a percentage. So maybe this check should be the first thing done in `parse_percent_rate()`?

ndev2 replied:
I think`tc_class_modify()` or `tc_qdisc_modify()` throw an error when no device is being specified. And the control goes from this code to the qdisc file (which parses the rate, class arguments, etc) so doesn't that already take care of the 'no device specification scenario' ? (just thinking out loud here)

ndev2 replied:
But even in that case, I guess this would serve as a safety check inside `parse_percent_rate()` which is necessary. But do let me know what you think. Thanks.

ndev2 commented:
Hi Cody, I've rebased the code with all the changes incorporated, and the `checkpatch.pl` checked, hope the code is good. Thanks.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -191,6 +191,61 @@ static const struct rate_suffix {
+	int ret;
+	double perc, rate_bits;
+
+	if (dev == NULL) {
```
Since we declare the space for the device name on the stack in `tc_qdisc_modify()`, `dev` will never be `NULL`. Instead, the test should be `if (!dev[0]) {`.

cjdoucette commented on tc/tc_util.c:
```diff
@@ -191,6 +191,61 @@ static const struct rate_suffix {
+
+	if (dev == NULL) {
+		fprintf(stderr, "No device specified; please specify device\n"
+);
```
Let's make this `"No device specified; specify device to rate limit by percentage\n"`.

Also, move the `);` onto the previous line.

ndev2 commented:
Rebased with the changes Cody, thanks. Do let me know if any other changes to be made, or should we  go ahead with the upstreaming process?

cjdoucette commented:
OK, it looks good to me now. I'm just going to have Michel look over the patch to see if he has any suggestions.

My main concern is still that we are passing the device name to all of the individual qdiscs, which is a dependency that didn't exist before. But the fact that the device name is a required parameter to `tc` helps our case, and we don't really have a way around it: there's no way in userspace to access what device is at the top of the hierarchy of the qdisc that you're using.

Right now, when the user does not specify a device name, the kernel returns an `ENODEV` error. In other words, it won't try to find what device the user should have specified. Based on those reasons, I think the patch will be OK, but we may have to argue for it.

ndev2 commented:
Ack! I look forward to it. I understand the way we handle the `dev` parameter is a little shaky, but the existing support looks like not much importance is give to the `dev` parameter anyways (as you mentioned), and given that, I think our solution works well, without stirring up the existing architecture as much.
